source/beamline/AMPVControl.cpp

```diff
@@ -672,6 +650,106 @@ void AMPVwStatusAndUnitConversionControl::setUnitConverters(AMAbstractUnitConver
 		emit setpointChanged(newSetpoint);
 }
 
-AMControlStatusCheckerDefault::~AMControlStatusCheckerDefault(){}
-AMControlStatusCheckerStopped::~AMControlStatusCheckerStopped(){}
-AMScaleAndOffsetUnitConverter::~AMScaleAndOffsetUnitConverter(){}
+/// ========================= implementation of AMReadOnlyWaveformBinningSinglePVControl =================
+AMReadOnlyWaveformBinningPVControl::~AMReadOnlyWaveformBinningPVControl(){}
+AMReadOnlyWaveformBinningPVControl::AMReadOnlyWaveformBinningPVControl(const QString &name, const QString &readPVname, int lowIndex, int highIndex, QObject *parent, const QString &description) :
+	AMReadOnlyPVControl(name, readPVname, parent, description)
+{
+	disconnect(readPV_, SIGNAL(valueChanged(double)), this, SIGNAL(valueChanged(double)));
+	attemptDouble_ = false;
+	setBinParameters(lowIndex, highIndex);
+	connect(readPV_, SIGNAL(valueChanged()), this, SLOT(onReadPVValueChanged()));
+}
+
+double AMReadOnlyWaveformBinningPVControl::value() const{
+	if(attemptDouble_)
+		return readPV_->binFloatingPointValues(lowIndex_, highIndex_);
+	return readPV_->binIntegerValues(lowIndex_, highIndex_);
+}
+
+void AMReadOnlyWaveformBinningPVControl::setBinParameters(int lowIndex, int highIndex){
+	lowIndex_ = lowIndex;
+	highIndex_ = highIndex;
+}
+
+void AMReadOnlyWaveformBinningPVControl::setAttemptDouble(bool attemptDouble){
+	attemptDouble_ = attemptDouble;
+}
+
+void AMReadOnlyWaveformBinningPVControl::onReadPVValueChanged(){
+	emit valueChanged(value());
+}
+
+/// ========================= implementation of AMWaveformBinningSinglePVControl =================
+AMWaveformBinningSinglePVControl::AMWaveformBinningSinglePVControl(const QString &name, const QString &PVName, int lowIndex, int highIndex, QObject *parent, const QString &description) :
+	AMSinglePVControl(name, PVName, parent)
+{
+	setDescription(description);
+	setAttemptDouble(false);
+	setBinParameters(lowIndex, highIndex);
+
+	// reconnect the valueChanged signal for readPV_
+	disconnect(readPV_, SIGNAL(valueChanged(double)), this, SIGNAL(valueChanged(double)));
+	connect(readPV_, SIGNAL(valueChanged()), this, SLOT(onReadPVValueChanged()));
+}
+
+double AMWaveformBinningSinglePVControl::value() const{
+	if(attemptDouble_)
+		return readPV_->binFloatingPointValues(lowIndex_, highIndex_);
+	return readPV_->binIntegerValues(lowIndex_, highIndex_);
+}
+
+bool AMWaveformBinningSinglePVControl::values(int size, int *outputValues) const{
+	if(attemptDouble_)
```

Comment:
Nevermind, I see what is happening now.  You'd probably be better off having just a single values and just cast the integers to doubles if the type isn't right.